### ISSUE TITLE
[ONNX] exported zero-dim constant type is wrong

### ISSUE THREAD

**Issue by dashesy** [pytorch/pytorch#26184]
## 🐛 Bug

```python
two = torch.tensor(2.0, dtype=torch.float32)
```

Should be exported as float32 but is exported as float64, which throws off this error in ORT:

> RuntimeError: [ONNXRuntimeError] : 1 : GENERAL ERROR : Load model from output/gr/logo/logo.onnx failed:Type Error: Type parameter (T) bound to different types (tensor(float) and tensor(double) in node (rand_node_name_330).

## To Reproduce

```python
        two = torch.tensor(2.0, dtype=torch.float32)
        half = torch.tensor(0.5, dtype=torch.float32)
        crop_size = torch.as_tensor(self.crop_size, dtype=torch.float32)
        top = ((target_height - crop_size) / two + half).floor()
        left = ((target_width - crop_size) / two + half).floor()
```

![Capture](https://user-images.githubusercontent.com/873905/64888453-ebce3200-d61f-11e9-8214-89e03d72e67f.JPG)

Div's constant should be `float32`

## Expected behavior

Export correct type for zero-dim constant

## Environment

PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.14.0

Python version: 2.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1070
Nvidia driver version: 418.40.04
cuDNN version: /usr/local/lib/libcudnn.so.5.1.10

Versions of relevant libraries:
[pip] mictorch==0.0.1
[pip] numpy==1.16.4
[pip] torch==1.2.0
[pip] torchvision==0.4.0
[conda] Could not collect

## Additional context

The workaround was to make it a single-dim constant, this is exported correctly

```python
        two = torch.tensor([2.0], dtype=torch.float32)
        half = torch.tensor([0.5], dtype=torch.float32)
        crop_size = torch.as_tensor(self.crop_size, dtype=torch.float32)
        top = ((target_height - crop_size) / two + half).floor()
        left = ((target_width - crop_size) / two + half).floor()
```

**Comment by neginraoof** [comment]
I'm unable to reproduce this problem. I tested your model with:
two = torch.tensor(2.0, dtype=torch.float32)
half = torch.tensor(0.5, dtype=torch.float32)
And didn't see any issues. Please give it a try.

**Comment by dashesy** [comment]
It reproduces for me

> RuntimeError: [ONNXRuntimeError] : 1 : GENERAL ERROR : Load model from output/gr/logo/logo.onnx failed:Type Error: Type parameter (T) bound to different types (tensor(float) and tensor(double) in node (rand_node_name_322).


I export in Python2, could that be the difference? lmk I can debug

**Comment by dashesy** [comment]
This is no issue in Python3, so must be Python2 related

**Comment by mrevow** [comment]
Thanks this helped me out. However I ran into the same issue with python 3

Python: 3.7.3
onnx ver 1.5.0
pytorch ver: 1.1.0
platform: Windows 10

The same workaround mentioned above (making single-dim constant,) by @dashesy worked

The issue did not occur for me when using pytorch 1.2.0

**Comment by willyd** [comment]
Same issue here:

python==3.8.0
onnx==1.8.0
torch==1.7.0
torchvision==0.8.1

**Comment by willyd** [comment]
Here is a small script that reproduces the error:
```python
import difflib

import onnx
import torch


@torch.jit.script
def add_with_zero_dim_tensor(x):
    if x.numel() <= 0:
        return torch.empty((1,), dtype=torch.float32)
    else:
        x_max = x.max()
        y = x_max + torch.tensor(1, dtype=torch.float32)
        return y * 2

@torch.jit.script
def add_with_one_dim_tensor(x):
    if x.numel() <= 0:
        return torch.empty((1,), dtype=torch.float32)
    else:
        x_max = x.max()
        y = x_max + torch.tensor([1], dtype=torch.float32)
        return y * 2

class AddWithZeroDimTensor(torch.nn.Module):

    def forward(self, x):
        return add_with_zero_dim_tensor(x)

class AddWithOneDimTensor(torch.nn.Module):

    def forward(self, x):
        return add_with_one_dim_tensor(x)

def trace_and_print(model, x):
    script = torch.jit.trace(model, x)
    print(script.graph)

def export_to_onnx_and_print(model, x):
    outname = f'{model.__class__.__name__}.onnx'
    torch.onnx.export(model, x, outname, input_names=['x'])
    return str(onnx.load(outname)).split('\n')

x = torch.ones((10,), dtype=torch.float32)

s1 = export_to_onnx_and_print(AddWithZeroDimTensor(), x)
s2 = export_to_onnx_and_print(AddWithOneDimTensor(), x)
print('\n'.join(difflib.Differ().compare(s1, s2)))
```
produces the following output:
```diff
  ir_version: 6
  producer_name: "pytorch"
  producer_version: "1.7"
  graph {
    node {
      input: "x"
      output: "1"
      name: "Shape_0"
      op_type: "Shape"
    }
    node {
      input: "1"
      output: "2"
      name: "ReduceProd_1"
      op_type: "ReduceProd"
      attribute {
        name: "keepdims"
        i: 0
        type: INT
      }
    }
    node {
      output: "3"
      name: "Constant_2"
      op_type: "Constant"
      attribute {
        name: "value"
        t {
          data_type: 7
          raw_data: "\000\000\000\000\000\000\000\000"
        }
        type: TENSOR
      }
    }
    node {
      input: "2"
      input: "3"
      output: "4"
      name: "Greater_3"
      op_type: "Greater"
    }
    node {
      input: "4"
      output: "5"
      name: "Not_4"
      op_type: "Not"
    }
    node {
      input: "5"
      output: "6"
      name: "If_5"
      op_type: "If"
      attribute {
        name: "then_branch"
        g {
          node {
            output: "7"
            name: "Constant_6"
            op_type: "Constant"
            attribute {
              name: "value"
              t {
                dims: 1
                data_type: 1
-               raw_data: "\240\\0+"
?                           ^^    ^

+               raw_data: "\000\000\000\000"
?                           ^^  +++  ^^^^^^

              }
              type: TENSOR
            }
          }
          name: "torch-jit-export1"
          output {
            name: "7"
            type {
              tensor_type {
                elem_type: 1
                shape {
                  dim {
                    dim_value: 1
                  }
                }
              }
            }
          }
        }
        type: GRAPH
      }
      attribute {
        name: "else_branch"
        g {
          node {
            input: "x"
            output: "8"
            name: "ReduceMax_7"
            op_type: "ReduceMax"
            attribute {
              name: "keepdims"
              i: 0
              type: INT
            }
          }
          node {
            output: "9"
            name: "Constant_8"
            op_type: "Constant"
            attribute {
              name: "value"
              t {
+               dims: 1
-               data_type: 11
?                           -

+               data_type: 1
-               raw_data: "\000\000\000\000\000\000\360?"
?                                    -----------------

+               raw_data: "\000\000\200?"
?                                   +

              }
              type: TENSOR
            }
          }
          node {
            input: "8"
            input: "9"
            output: "10"
            name: "Add_9"
            op_type: "Add"
          }
          node {
            output: "11"
            name: "Constant_10"
            op_type: "Constant"
            attribute {
              name: "value"
              t {
-               data_type: 11
?                           -

+               data_type: 1
-               raw_data: "\000\000\000\000\000\000\000@"
?                                      ----------------

+               raw_data: "\000\000\000@"
              }
              type: TENSOR
            }
          }
          node {
            input: "10"
            input: "11"
            output: "12"
            name: "Mul_11"
            op_type: "Mul"
          }
          name: "torch-jit-export2"
          output {
            name: "12"
            type {
              tensor_type {
-               elem_type: 11
?                           -

+               elem_type: 1
              }
            }
          }
        }
        type: GRAPH
      }
    }
    name: "torch-jit-export"
    input {
      name: "x"
      type {
        tensor_type {
          elem_type: 1
          shape {
            dim {
              dim_value: 10
            }
          }
        }
      }
    }
    output {
      name: "6"
      type {
        tensor_type {
          elem_type: 1
          shape {
+           dim {
+             dim_value: 1
+           }
          }
        }
      }
    }
  }
  opset_import {
    version: 9
  }
```
In the first case the datatype is set to `11` which is float64 and in the second case the datatype is set to `1` which is float32.

**Comment by garymm** [comment]
Seems this is fixed in PyTorch 1.10.0.
